### PR TITLE
Use json-iterator instead of ugorji for JSON.

### LINKED ISSUE #36120: Remove ugorji codecgen in 1.6
Now that we are moving to pure protobuf in etcd and for internal channels, we no longer need ugorji codec generation to make JSON fast.  We can use the ugorji marshaler / unmarshaler (it has minor perf benefits) or just revert to using golang JSON serialization.  The ratio of total cluster traffic using JSON should be small enough that the reduction in code size, generated code, and build logic is worth the tradeoff.

@kubernetes/sig-scalability

### PR DESCRIPTION
**Pull request by thockin**
@smarterclayton @wojtek-t

Fixes #36120
xref #18762

```release-note
Switch JSON marshal/unmarshal to json-iterator library.  Performance should be close to previous with no generated code.
```

### CODE REVIEW

**Comment by thockin** [comment]
This has hacked up godeps, but other than that - it actually passes tests now!  Do we have a relevant benchmark?  @smarterclayton, you had other concerns?

**Comment by thockin** [comment]
FWIW by the micro benchmarks, it shows Marshal 2x faster than stdlib and Unmarshal 3x or more faster.

**Comment by smarterclayton** [comment]
Other than custom marshallers, the registration for custom marshallers
(which assumed package short name + type, which is not unique enough), and
the possibility that the unsafe code is really unsafe, no.

**Comment by smarterclayton** [comment]
Do we need to do an aggressive fuzzing pass on this?

**Comment by thockin** [comment]
More fuzzing is more better.  I added a ton of tests to the json-iterator repo to fuzz just about every combination of types and structs and tags I could think up.  Many a direct result of making these tests pass.  We have the serialization test already - can we expand on that?  Do the same thing, but for our own complex types?

Or we could fuzz at the input level - generate random structs, write to files, run the round-trip test against them.

**Comment by smarterclayton** [comment]
I'm wondering if we should be JSON fuzzing (since this is a net new custom
parser).  Not just well formed input.

**Comment by thockin** [comment]
Ahh, yeah, probably, that too.

**Comment by thockin** [comment]
I added random-input fuzzing to json-iterator which, no surprise, caught some bugs.  They are working on those bugs, mostly fixed, one last one to do it sounds like.

https://github.com/json-iterator/go/blob/master/unmarshal_input_test.go

**Comment by smarterclayton** [comment]
Alignment bugs or unsafe memory access bugs?

**Comment by thockin** [comment]
Just parsing bugs - allowed invalid input when it shouldn't have.  To test alignment would require deeper understanding of their internals.

**Comment by thockin** [comment]
@smarterclayton new fun failure.  Apparently type `meta.Table` currently passes roundrip ONLY because ugorji behaves differently from stdlib when unmarshaling a number into `interface{}`.  Apparently it tries to unmarshal into a uint64 if it can, while both stdlib and json-iterator go straight for float64.

That results in the test fails evident here.

**Comment by thockin** [comment]
ping @smarterclayton - I'm still hoping beyond hope that some parts of the rebase hell can be fixed before 1.8 lockdown.  I don't know anything about meta.Table, need guidance.

**Comment by thockin** [comment]
Rebased, but the Table issue still exists.  I tried forcing it to decode how we want, but no luck yet.

**Comment by thockin** [comment]
yeah, @smarterclayton - Table is busted wrt RoundTrip testing.  I can't know whether it fuzzed an int64 or a uint64.  What is Table about?  Does it really need to be interface{} ?  Can we constrain it in fuzzing?

**Comment by thockin** [comment]
OK, I am almost there.  jsoniter has a type-decode override, which works ALMOST how we need...

**Comment by thockin** [comment]
OK, it seems the bug in json-iterator is fixed.  Show me green, baby....

**Comment by thockin** [comment]
Interestingly, this seems to fix the map issue documented in the fuzzer for Table.

**Comment by thockin** [comment]
While other PRs queue up, thoughts on this for 1.8?  If not 1.8, early in 1.9?

**Comment by wojtek-t** [comment]
I would really like to see this in for 1.8. I'm fine with the PR as is (we should probably rename commits to remove WIP from their names).

@smarterclayton - do you have anything specific on your mind before we can merge it?

**Comment by thockin** [comment]
I'll reorg commits and rebase

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/fuzzer/fuzzer.go [hunk 1]
```diff
@@ -145,15 +145,11 @@ func v1alpha1FuzzerFuncs(codecs runtimeserializer.CodecFactory) []interface{} {
...
-					// maps roundtrip as map[interface{}]interface{}, but the json codec cannot encode that
-					// TODO: get maps to roundtrip properly
-					/*
-						x := map[string]interface{}{}
-						for j := c.Intn(10) + 1; j >= 0; j-- {
-							x[c.RandString()] = c.RandString()
-						}
-						r.Cells[i] = x
-					*/
+					x := map[string]interface{}{}
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/fuzzer/fuzzer.go]: yay

**Comment by smarterclayton** [comment]
If we put this in now, do we have enough soak for release?  Can we revert easily?

This is better than ugorji other than the risk of a potentially uncaught security issue.  One option - we could stick a build tag on the generated code and disable it all at once?

**Comment by thockin** [comment]
Rebased and commits cleaned up.  PTAL.

@smarterclayton not sure what you mean.  We could make this a flag or something, but that doesn't get rid of all the types.generated.go conflicts.  I am anxious about this, but the only way to really proceed is to get miles on it.

**Comment by thockin** [comment]
I'll re-vendor the newest iterator, too

**Comment by smarterclayton** [comment]
No, I mean, we merge this (and have ugorji gone) - can we disabled it quickly and easily if we decide it needs more soak time?

**Comment by smarterclayton** [comment]
Ok, with the split commits it's more obvious how we could disable this in a quick patch.

I'm ok with this merging for 1.8 as long as we keep an eye on it and are open to disable / revert if necessary.

**Comment by thockin** [comment]
Updated the ugorji dep.  Come on CI... show me green.

**Comment by thockin** [comment]
I have re-titled this and added a release note.

**Comment by wojtek-t** [comment]
> I'm ok with this merging for 1.8 as long as we keep an eye on it and are open to disable / revert if necessary.

+1

/lgtm

**Comment by thockin** [comment]
rebased, only files touched were types.generated.go

/lgtm

**Comment by thockin** [comment]
/retest

FWIW, upstream json-iterator tagged this hash as v1.0.0

**Comment by thockin** [comment]
/retest 

again

**Comment by liggitt** [comment]
You have made me a happy man.

**Comment by thockin** [comment]
HOLY SMOKE, it's green!! Come one merge queue!!

**Comment by thockin** [comment]
Next I will aitm at godeps and the rest of the generated code.  That other PR has been open for OVER A YEAR.  That will leave protobuf as the primary offender.  And boy is it offensive.

**Comment by smarterclayton** [comment]
Psh, protobuf is fast.

**Comment by thockin** [comment]
/retest

Son of a ...

**Comment by thockin** [comment]
/retest

I'm gonna kill someone.

**Comment by thockin** [comment]
HOLY CRAP WHY WONT THE TESTS PASS HOW DO WE GET ANYTHING DONE?

/retesth

**Comment by thockin** [comment]
I am at wits end with the tests.

/retest

**Comment by dims** [comment]
@thockin known problem i think https://github.com/kubernetes/kubernetes/issues/51692

**Comment by thockin** [comment]
Would anyone object to bumping the prio up?  This will eliminate a lot of rebases that people have to do in the coming days..

**Comment by liggitt** [comment]
> Would anyone object to bumping the prio up? This will eliminate a lot of rebases that people have to do in the coming days..

not at all... I like the relative transparency of the `queue/multiple-rebases` label

**Comment by liggitt** [comment]
> This will eliminate a lot of rebases that people have to do in the coming days..

though the gzipped proto manifest still guarantees a conflict on any API change in the same group. c'est la vie

**Comment by thockin** [comment]
I have a different PR for removing conversion, deepcopy, and defaults.  I couldn't crack protobuf, and opted to leave it for "phase 2" but I can't get phase 1 merged. :(

**Comment by liggitt** [comment]
> thockin commented 35 seconds ago 
> I can't get phase 1 merged. :(

> k8s-merge-robot commented just now 
> @thockin PR needs rebase

you have angered the bot

**Comment by thockin** [comment]
I take the beating so our contributors don't have to.

**Comment by wojtek-t** [comment]
+1 for bumping priority

The problem is that all the test rekicks were because tests didn't pass when trying to reach the queue, not during presubmit while merging...

**Comment by smarterclayton** [comment]
I'm pretty sure everyone else takes the beatings and just accepts them as
their lot in life for contributing to open source.

Who needs angry users when the bot has it out for you?

**Comment by thockin** [comment]
I am rebasing faster than I can complete a test run.

**Comment by thockin** [comment]
@k8s-ci-robot The hell I can't.  Hold my beer.

**Comment by fejta** [comment]
🍿🍿🍿😂😂😂

**Comment by sttts** [comment]
Go submit-queue, go!

**Comment by sttts** [comment]
🎉  Wow!  🎉

**Comment by wojtek-t** [comment]
Nice!

**Comment by luxas** [comment]
Yay!
(Yeah, this is what it looks like to get a larger PR merged :(, know it too well)

**Comment by thockin** [comment]
w00t!   no rollback, no rollback, no rollback...

**Comment by justinsb** [comment]
I want to see the impact on the binary size... and IIUC it is every binary (e.g. kubectl)

**Comment by thockin** [comment]
@justinsb - good point, also build time.  Deleted 190KLOC :)

**Comment by ugorji** [comment]
In case the issues you see are caused by the json-iterator merge ... 

FYI: See ugorji/go@54210f4e076c57f351166f0ed60e67d3fca57a36

Substantial performance updates that bring the performance without code-generation in line with what you get from json-iterator. we did this while only using 'unsafe' in very targeted places, with a build tag to switch to 'no unsafe' mode. See helper_unsafe.go to see the only functions where we used unsafe, with the equivalent 'safe' versions, to give you some peace of mind.

Results are attached, but I will reproduce the benchcmp output comparing results without codecgen and then with codecgen, here for convenience. These were done using the benchmark suite in https://github.com/ugorji/go-codec-bench , which uses a recursive embedding of a large detailed struct. The numbers speak for themselves. It might be worth going reverting sooner than later.

```sh
   go test -bench "_(Json|Cbor)" -benchmem -bd 2 -tags "x" > no-codecgen.bench.out.txt
   go test -bench "_(Json|Cbor)" -benchmem -bd 2 -tags "x codecgen" > with-codecgen.bench.out.txt
   benchcmp no-codecgen.bench.out.txt with-codecgen.bench.out.txt > benchcmp.out.txt
```

Results:
```
benchmark                          old ns/op     new ns/op     delta
Benchmark__Cbor_______Encode-8     69672         37051         -46.82%
Benchmark__Json_______Encode-8     115274        87775         -23.86%
Benchmark__JsonIter___Encode-8     133558        133635        +0.06%
Benchmark__Std_Json___Encode-8     165240        157007        -4.98%
Benchmark__Cbor_______Decode-8     164396        131192        -20.20%
Benchmark__Json_______Decode-8     207014        156589        -24.36%
Benchmark__JsonIter___Decode-8     209418        208993        -0.20%
Benchmark__Std_Json___Decode-8     553650        557783        +0.75%

benchmark                          old allocs     new allocs     delta
Benchmark__Cbor_______Encode-8     32             7              -78.12%
Benchmark__Json_______Encode-8     45             20             -55.56%
Benchmark__JsonIter___Encode-8     1013           1013           +0.00%
Benchmark__Std_Json___Encode-8     560            560            +0.00%
Benchmark__Cbor_______Decode-8     713            697            -2.24%
Benchmark__Json_______Decode-8     756            736            -2.65%
Benchmark__JsonIter___Decode-8     2244           2244           +0.00%
Benchmark__Std_Json___Decode-8     1967           1967           +0.00%

benchmark                          old bytes     new bytes     delta
Benchmark__Cbor_______Encode-8     5529          2320          -58.04%
Benchmark__Json_______Encode-8     6290          3072          -51.16%
Benchmark__JsonIter___Encode-8     40496         40496         +0.00%
Benchmark__Std_Json___Encode-8     65248         65248         +0.00%
Benchmark__Cbor_______Decode-8     34384         34104         -0.81%
Benchmark__Json_______Decode-8     42976         42608         -0.86%
Benchmark__JsonIter___Decode-8     60128         60128         +0.00%
Benchmark__Std_Json___Decode-8     58320         58320         +0.00%
```

[benchcmp.out.txt](https://github.com/kubernetes/kubernetes/files/1312517/benchcmp.out.txt)
[no-codecgen.bench.out.txt](https://github.com/kubernetes/kubernetes/files/1312515/no-codecgen.bench.out.txt)
[with-codecgen.bench.out.txt](https://github.com/kubernetes/kubernetes/files/1312516/with-codecgen.bench.out.txt)

**Pull request merged**